File: Difficulty: Easy/Largest subarray of 0's and 1's/largest-subarray-of-0s-and-1s.java
//{ Driver Code Starts
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        // Create BufferedReader for efficient input reading
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        // Read number of test cases
        int T = Integer.parseInt(br.readLine());

        // Process each test case
        while (T-- > 0) {
            // Read the entire line containing the array elements
            String line = br.readLine();

            // Split the line into an array of strings, then parse them as integers
            String[] tokens = line.split("\\s+");
            int[] a = new int[tokens.length];
            for (int i = 0; i < tokens.length; i++) {
                a[i] = Integer.parseInt(tokens[i]);
            }

            // Create the Solution object
            Solution obj = new Solution();

            // Call maxLen function and print the result
            System.out.println(obj.maxLen(a));
        }
    }
}
// } Driver Code Ends


class Solution {
    public int maxLen(int[] arr) {
        // Your code here
        int ans = 0, sum = 0;
        Map<Integer, Integer> map = new HashMap<>();
        map.put(0, -1);
        
        for(int i=0; i<arr.length; i++){
            sum += arr[i] == 1 ? 1 : -1;
            
            if(map.containsKey(sum)){
                ans = Math.max(ans, i-map.get(sum));
            }else{
                map.put(sum, i);
            }
        }
        return ans;
    }
}

File: Difficulty: Easy/Largest subarray of 0's and 1's/largest-subarray-of-0s-and-1s.py
class Solution:
    def maxLen(self, arr):
        # code here
        n = len(arr)
        
        prefix_sum = 0
        max_len = 0
        
        mp = {}
        mp[0] = -1
        
        for i in range(n):
            if arr[i] == 0:
                prefix_sum += -1
            else:
                prefix_sum += 1
                
            if prefix_sum in mp:
                length = i - mp[prefix_sum]
                if length > max_len:
                    max_len = length
            else:
                mp[prefix_sum] = i
                
        return max_len
            